lib/sdk/storage.js

```diff
@@ -47,26 +46,40 @@ async function getAsset(assetReference, directory, name, disableDownload) {
         }
         console.log("WORKER_TEST_MODE: asset path:", filePath);
         return new Asset(assetReference, directory, name);
+    }
 
-    } else {
-        const source = new Asset(assetReference, directory, name);
+    const source = new Asset(assetReference, directory, name);
+    if (isDataUrlAndValid(assetReference.url)) {
+        console.log("creating asset from data url:", source.path);
+        await datauri.download(assetReference, source.path);
+        if(!fileExistsAndIsNotEmpty(source.path)){
+            console.log(`file ${source.path} does not exist after datauri write to local fs`);
+            throw new SourceUnsupportedError(`Invalid or missing local file ${source.path}`);
+        }
         if (disableDownload) {
             console.log(`Skipping source file download for ${AssetComputeLogUtils.redactUrl(source.url)}`);
-        } else {
-            // TODO: error handling: errors catched & thrown in http.download() and then catched again in worker.js ???
-            // download http/https url into file
-            const protocol = URL.parse(assetReference.url).protocol;
-            if (protocol === DATA_PROTOCOL) {
-                console.log("creating asset from data url:", source.path);
-                await datauri.download(assetReference, source.path);
-            } else {
-                console.log("downloading asset:", source.path);
-                await http.download(assetReference, source.path);
-            }
+            const preSignedUrl = await datauri.getPreSignedUrl(source.path);
+            console.log(`Uploaded data URI content to storage and generated presigned url`);
+            assetReference.url = preSignedUrl;
+            return new Asset(assetReference, directory, name);
         }
+    } else if (!disableDownload) {
+        console.log("downloading asset:", source.path);
+        await http.download(assetReference, source.path);
+    }
+
+    return source;
+}
 
-        return source;
+// function to verify if a input is a data url
+// throw SourceUnsupportedError if it is a data url but invalid
+function isDataUrlAndValid(input) {
+    if (URL.parse(input).protocol !== 'data:') {
```

Comment:
`URL.parse` is deprecated. Is there a specific reason we need to check for protocol and return false, instead of `validDataUrl` check.

Reply:
+1

Reply:
What do we get from knowing the protocol is valid if afterwards the URL is malformed? We won't be able to use it (having a more specific error doesn't count as argument :) ).

Also I don't understand `if we only check if it is valid, then when is data protocol but malformed, we will treat it as http`: if we see a malformed URI, be it data, HTTPS or whichever protocol, we should stop processing and throw.

Reply:
For answering this `Is there a specific reason we need to check for protocol and return false, instead of validDataUrl check.`:
what I meant for this is `if we only check if it is valid, then when is data protocol but malformed, we will treat it as http`
ex. Assuming we only check if datauri is valid, and we have url = `data:image/a/bc;base64,xxxx` (where mimeType is malformed or something else). Since we consider it as not valid datauri, we stop and throw.
But consider when url = `https://example.com`, it is a valid http url but not a valid datauri. In this case we don't want to stop and throw. Instead we want it to be checked in later statement to see if it is valid http url.

So the reason of checking the protocol is as described above. Since we do not know if url is datauri or http in the first place, tp achieve this statement for `if we see a malformed URI, be it data, HTTPS or whichever protocol, we should stop processing and throw.`, checking the protocol then decide if it is valid for the protocol make sense.

Feel free to correct me if I misunderstand the original comments. @pheenomenon @tmathern

And the reason of naming is because the function does not only check if it is valid, but also check the protocol. But maybe we can have a better naming that explains what the func is checking than what I have now, haha @jdelbick

Reply:
use URL constructor instead

Reply:
To make the code easier to read I guess I will have a general validateUrl function and do it once at the top instead.
The function does the actual valid check, and in later condition we only checks the protocol to decide the action.

Reply:
You could write a check like this (function names may be wrong, I didn't check them):
```
if(isValidHttpsUri()){
} else if (isValidDataUri()){
} else {
// any other case
}
```

If you write the `if` condition this way, you check validity+protocol at the same time. I don't think you need to explicitly separate protocol check from URI check.

Concretely, the line https://github.com/adobe/asset-compute-pipeline/blob/8ff0a971fe7d5a955084295efb75932bb045748b/lib/sdk/storage.js#L52 could become the library function call:
```
if (isValidDataUri(assetReference.url))
```
Then getting the exception to throw needs changes n the else conditions (https://github.com/adobe/asset-compute-pipeline/blob/8ff0a971fe7d5a955084295efb75932bb045748b/lib/sdk/storage.js#L66-L71): add a verification the HTTPS URL is valid, then download. 

The idea is that you don't need additional validation functions, use those that are already provided/dependencies of the pipeline.